org.adempiere.base/src/org/compiere/model/PO.java

```diff
@@ -1884,31 +1959,55 @@ private void setKeyInfo()
 		}
 		//	Set FKs
 		int size = columnNames.size();
-		if (size == 0)
-			throw new IllegalStateException("No PK nor FK - " + p_info.getTableName());
-		m_IDs = new Object[size];
-		m_KeyColumns = new String[size];
-		for (int i = 0; i < size; i++)
+		if (size > 0)
 		{
-			m_KeyColumns[i] = (String)columnNames.get(i);
-			if (m_KeyColumns[i].endsWith("_ID"))
+			m_IDs = new Object[size];
+			m_KeyColumns = new String[size];
+			for (int i = 0; i < size; i++)
 			{
-				Integer ii = null;
-				try
+				m_KeyColumns[i] = (String)columnNames.get(i);
+				if (m_KeyColumns[i].endsWith("_ID"))
 				{
-					ii = (Integer)get_Value(m_KeyColumns[i]);
+					Integer ii = null;
+					try
+					{
+						ii = (Integer)get_Value(m_KeyColumns[i]);
+					}
+					catch (Exception e)
+					{
+						log.log(Level.SEVERE, "", e);
+					}
+					if (ii != null)
+						m_IDs[i] = ii;
 				}
-				catch (Exception e)
+				else
+					m_IDs[i] = get_Value(m_KeyColumns[i]);
+				if (log.isLoggable(Level.FINEST)) log.finest("(FK) " + m_KeyColumns[i] + "=" + m_IDs[i]);
+			}
+		}
+
+		if (m_KeyColumns == null || m_KeyColumns.length == 0)
+		{
```

Comment:
This is inconsistent with GridTable.getWhereClause. At GridTable.getWhereClause _UU take precedence over isColumnParent (i.e multi key) and it is the other way round here.

Reply:
With the new setKeyInfo code, there's one troubling difference here - MTable.getKeyColumns and PO.get_KeyColumns could be different. Thus breaking a very well known API contract.

Reply:
> With the new setKeyInfo code, there's one significant side effect here - MTable.getKeyColumns and PO.get_KeyColumns could be different. Thus breaking a very well known API contract.

Readu, changed back GridTable to keep behavior same as PO